src/lib/protocols/slp.c

```diff
@@ -0,0 +1,286 @@
+static void ndpi_dissect_slp_v2(struct ndpi_detection_module_struct *ndpi_struct,
+                                struct ndpi_flow_struct *flow)
+{
+  struct ndpi_packet_struct const * const packet = &ndpi_struct->packet;
+  struct slp_hdr_v2 const * const hdr = (struct slp_hdr_v2 *)&packet->payload[0];
+  int url_offset = -1; // Can be either an offset to <URL String> or <URL Entry>
+  int url_length_offset = -1; // length of <URL String>
+  int url_entry_count_offset = -1; // amount of <URL Entry>'s
+
+  switch (hdr->function_id) {
+    case FID_SrvRply:
+      url_entry_count_offset = 2;
+      url_offset = 4;
+      break;
+    case FID_SrvReg:
+      url_offset = 3; // contains always 1 <URL Entry>
+      break;
+    case FID_DAAdvert:
+      url_length_offset = 6;
+      url_offset = 8;
+      break;
+    case FID_SAAdvert:
+      url_length_offset = 0;
+      url_offset = 2;
+      break;
+    case FID_AttrRqst:
+      url_length_offset = 4;
+      url_offset = 6;
+      break;
+    case FID_SrvDeReg:
+      url_offset = 7; // contains always 1 <URL Entry>
+      break;
+  }
+
+  if (url_offset >= 0) {
+    uint16_t url_length_or_count = 0;
+
+    if (url_length_offset > 0 && packet->payload_packet_len > sizeof(*hdr) + url_length_offset + 2) {
+      // <URL String>
+      url_length_or_count = ntohs(*(uint16_t *)&packet->payload[sizeof(*hdr) + url_length_offset]);
+      if (packet->payload_packet_len > sizeof(*hdr) + url_offset + 2 + url_length_or_count) {
+        size_t len = ndpi_min(sizeof(flow->protos.slp.url[0]) - 1, url_length_or_count);
+        flow->protos.slp.url_count = 1;
+        strncpy(flow->protos.slp.url[0], (char *)&packet->payload[sizeof(*hdr) + url_offset + 2], len);
+        flow->protos.slp.url[0][len] = 0;
+      }
+    } else if (url_entry_count_offset > 0 && packet->payload_packet_len > sizeof(*hdr) + url_entry_count_offset + 2) {
+      // TODO: <URL Entry>'s
+      // Provide a PCAP file.
+      //url_length_or_count = ntohs(*(uint16_t *)&packet->payload[sizeof(*hdr) + url_entry_count_offset]);
+    } else if (packet->payload_packet_len > sizeof(*hdr) + url_offset + 2) {
+      url_length_or_count = ntohs(*(uint16_t *)&packet->payload[sizeof(*hdr) + url_offset]); // FID_SrvReg or FID_SrvDeReg
+      if (packet->payload_packet_len > sizeof(*hdr) + url_offset + 2 + url_length_or_count) {
+        size_t len = ndpi_min(sizeof(flow->protos.slp.url[0]) - 1, url_length_or_count);
+        flow->protos.slp.url_count = 1;
+        strncpy(flow->protos.slp.url[0], (char *)&packet->payload[sizeof(*hdr) + url_offset + 2], len);
+        flow->protos.slp.url[0][len] = 0;
+      }
+    }
+  }
+}
+
+static void ndpi_search_slp(struct ndpi_detection_module_struct *ndpi_struct,
+                            struct ndpi_flow_struct *flow)
+{
+  struct ndpi_packet_struct const * const packet = &ndpi_struct->packet;
+
+  NDPI_LOG_DBG(ndpi_struct, "search Service Location Protocol\n");
+
+  if (packet->payload_packet_len < 1) {
```

Comment:
Is this check really necessary? We should always have `packet->payload_packet_len >= 1`, because of `NDPI_SELECTION_BITMASK_PROTOCOL_V4_V6_TCP_OR_UDP_WITH_PAYLOAD_WITHOUT_RETRANSMISSION`

Reply:
True. Will remove.